### PR TITLE
fix: Load events to fix program rule variables calculation [DHIS2-17951]

### PR DESCRIPTION
**Pull request by enricocolasante**
`DefaultProgramRuleService` was wrongly trying to get all the events of an enrollment from the `TrackerPreheat` but they are not loaded in there and they need to be retrieved using `EventService`. Hence, the calculation of the correct dataValue based on the previous event was not working as previous event was not found unless it was present in the payload.

This change exposed some issues that we had in tests as some users didn't have org units in either the search scope and the capture scope and this is an invalid configuration that was getting validated by the event exporter.

Furthermore, when we inject different users in different tests, we need to make sure that we know which user is injected if we do not explicitly do so in the test. So in tests where we inject different users in tests we need to add a `@BeforeEach` method to inject the default user.

### CODE REVIEW

File: dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/imports/programrule/DefaultProgramRuleService.java [hunk 1]
```diff
@@ -202,10 +212,22 @@ private List<TrackedEntityAttributeValue> getAttributes(
...
-            .filter(e -> e.getEnrollment().getUid().equals(enrollmentUid))
-            .filter(e -> bundle.findEventByUid(e.getUid()).isEmpty());
+    Stream<Event> events;
+    try {
+      events =
+          eventService
+              .getEvents(
+                  EventOperationParams.builder()
+                      .eventParams(EventParams.TRUE)
+                      .orgUnitMode(ACCESSIBLE)
```
Comment (teleivo) [inline dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/imports/programrule/DefaultProgramRuleService.java:222]: Who is the currentUser/who set it when this is executed?

File: dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/imports/programrule/DefaultProgramRuleService.java [hunk 2]
```diff
@@ -202,10 +212,22 @@ private List<TrackedEntityAttributeValue> getAttributes(
...
+                      .enrollments(Set.of(enrollmentUid))
+                      .build())
+              .stream()
+              .filter(e -> e.getEnrollment().getUid().equals(enrollmentUid))
```
Comment (teleivo) [inline dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/imports/programrule/DefaultProgramRuleService.java:226]: Why is this necessary? 🤔 should we not only get events for this enrollment as you put that into the params?

File: dhis-2/dhis-test-integration/src/test/java/org/hisp/dhis/tracker/export/event/EventChangeLogServiceTest.java [hunk 3]
```diff
@@ -157,8 +164,6 @@ void shouldReturnChangeLogsWhenDataValueIsCreated() throws NotFoundException, Fo
   @Test
   void shouldReturnChangeLogsWhenDataValueIsDeleted()
       throws NotFoundException, IOException, ForbiddenException {
-    testAsUser("M5zQapPyTZI");
```
Comment (teleivo) [inline dhis-2/dhis-test-integration/src/test/java/org/hisp/dhis/tracker/export/event/EventChangeLogServiceTest.java:160]: Is this method still needed? should we remove it

File: dhis-2/dhis-test-integration/src/test/java/org/hisp/dhis/tracker/imports/programrule/ProgramRuleAssignActionTest.java [hunk 4]
```diff
@@ -119,9 +130,62 @@ void shouldNotImportWithWarningWhenAttributeWithSameValueIsAssignedByAssignRule(
...
+    assertHasOnlyWarnings(importReport, E1308);
+
+    Event event = manager.get(Event.class, "D9PbzJY8bZZ");
+
+    List<String> eventDataValues =
+        event.getEventDataValues().stream()
+            .filter(dv -> dv.getDataElement().equals("DATAEL00002"))
+            .map(dv -> dv.getValue())
+            .toList();
+    Assertions.assertContainsOnly(List.of("THIRD"), eventDataValues);
```
Comment (teleivo) [inline dhis-2/dhis-test-integration/src/test/java/org/hisp/dhis/tracker/imports/programrule/ProgramRuleAssignActionTest.java:156]: add static import

File: dhis-2/dhis-test-integration/src/test/java/org/hisp/dhis/tracker/imports/programrule/ProgramRuleAssignActionTest.java [hunk 5]
```diff
@@ -119,9 +130,62 @@ void shouldNotImportWithWarningWhenAttributeWithSameValueIsAssignedByAssignRule(
...
+
+    ImportReport importReport = trackerImportService.importTracker(params, trackerObjects);
+    assertHasOnlyWarnings(importReport, E1308);
+
+    Event event = manager.get(Event.class, "D9PbzJY8bZZ");
+
+    List<String> eventDataValues =
+        event.getEventDataValues().stream()
+            .filter(dv -> dv.getDataElement().equals("DATAEL00002"))
+            .map(dv -> dv.getValue())
```
Comment (teleivo) [inline dhis-2/dhis-test-integration/src/test/java/org/hisp/dhis/tracker/imports/programrule/ProgramRuleAssignActionTest.java:154]: a method reference should work right?

File: dhis-2/dhis-test-integration/src/test/resources/tracker/programrule/event_with_last_date.json [hunk 6]
```diff
@@ -0,0 +1,79 @@
+{
+  "importMode": "COMMIT",
```
Comment (teleivo) [inline dhis-2/dhis-test-integration/src/test/resources/tracker/programrule/event_with_last_date.json:2]: This file is parsed into `TrackerObjects` so should it not only contain

```java
public class TrackerObjects implements Serializable {
  @JsonProperty @Builder.Default
  private final List<TrackedEntity> trackedEntities = new ArrayList<>();
  @JsonProperty @Builder.Default private final List<Enrollment> enrollments = new ArrayList<>();
  @JsonProperty @Builder.Default private final List<Event> events = new ArrayList<>();
  @JsonProperty @Builder.Default private final List<Relationship> relationships = new ArrayList<>();
}
```

?

This would make our files easier to read as we mostly put in fields that are ignored anyway like `createdAt`, ... 😅

File: dhis-2/dhis-test-integration/src/test/java/org/hisp/dhis/tracker/imports/programrule/ProgramRuleAssignActionTest.java [hunk 7]
```diff
@@ -119,9 +130,62 @@ void shouldNotImportWithWarningWhenAttributeWithSameValueIsAssignedByAssignRule(
     assertHasOnlyWarnings(importReport, E1310);
   }
 
+  @Test
+  void shouldImportEventWithLastDateAndCorrectlyAssignPreviousEventDataValue() throws IOException {
```
Comment (teleivo) [inline dhis-2/dhis-test-integration/src/test/java/org/hisp/dhis/tracker/imports/programrule/ProgramRuleAssignActionTest.java:134]: It's hard to see from this test what we are testing and what state is changing in the process 😅 

Why is it that you expect THIRD here and SECOND in the next test? I would like to see the difference in input/config that causes that inside the test.

File: dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/imports/programrule/DefaultProgramRuleService.java (hunk 1 above)
Comment (enricocolasante) [reply dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/imports/programrule/DefaultProgramRuleService.java:222]: This is using current user.
Exporters uses current user to perform validations.

File: dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/imports/programrule/DefaultProgramRuleService.java (hunk 2 above)
Comment (enricocolasante) [reply dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/imports/programrule/DefaultProgramRuleService.java:226]: It is not, copy and paste code :grimacing:

File: dhis-2/dhis-test-integration/src/test/java/org/hisp/dhis/tracker/export/event/EventChangeLogServiceTest.java (hunk 3 above)
Comment (enricocolasante) [reply dhis-2/dhis-test-integration/src/test/java/org/hisp/dhis/tracker/export/event/EventChangeLogServiceTest.java:160]: it is still used to inject other users in the test

File: dhis-2/dhis-test-integration/src/test/java/org/hisp/dhis/tracker/imports/programrule/ProgramRuleAssignActionTest.java (hunk 5 above)
Comment (enricocolasante) [reply dhis-2/dhis-test-integration/src/test/java/org/hisp/dhis/tracker/imports/programrule/ProgramRuleAssignActionTest.java:154]: Yes, that should have been caught by Sonar :thinking:

File: dhis-2/dhis-test-integration/src/test/java/org/hisp/dhis/tracker/imports/programrule/ProgramRuleAssignActionTest.java (hunk 4 above)
Comment (enricocolasante) [reply dhis-2/dhis-test-integration/src/test/java/org/hisp/dhis/tracker/imports/programrule/ProgramRuleAssignActionTest.java:156]: Done

File: dhis-2/dhis-test-integration/src/test/resources/tracker/programrule/event_with_last_date.json (hunk 6 above)
Comment (enricocolasante) [reply dhis-2/dhis-test-integration/src/test/resources/tracker/programrule/event_with_last_date.json:2]: Done

File: dhis-2/dhis-test-integration/src/test/java/org/hisp/dhis/tracker/imports/programrule/ProgramRuleAssignActionTest.java (hunk 7 above)
Comment (enricocolasante) [reply dhis-2/dhis-test-integration/src/test/java/org/hisp/dhis/tracker/imports/programrule/ProgramRuleAssignActionTest.java:134]: Done

File: dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/imports/programrule/DefaultProgramRuleService.java (hunk 1 above)
Comment (teleivo) [reply dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/imports/programrule/DefaultProgramRuleService.java:222]: but the program rules run as part of the import job which sets the current user to the user making the HTTP request, right?

**Review by teleivo — APPROVED** [review]

File: dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/imports/programrule/DefaultProgramRuleService.java (hunk 1 above)
Comment (enricocolasante) [reply dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/imports/programrule/DefaultProgramRuleService.java:222]: Yes yes, program rules run as part of import job, so they are using the same user.
